### ISSUE TITLE
[inductor] Multiple randint calls cause inconsistent RNG results between eager and compiled mode after fixing the rng seed

### ISSUE THREAD

**Issue by maybeLee** [pytorch/pytorch#175156]
### 🐛 Describe the bug

To reproduce:
```python
import torch
import torch._dynamo
torch._inductor.config.fallback_random = True

def fn():
    torch.manual_seed(0)
    out = torch.randint(0, 100, (4,1), dtype=torch.int64)  # Shape and integer range are not important.
    other = torch.randint(0, 100, (2,1), dtype=torch.int64) # No inconsistency if remove this line, although the value won't be returned.
    return out
    
cfunc = torch.compile(fn, backend='inductor')
torch.manual_seed(0)
out1 = fn()
torch.manual_seed(0)
out2 = cfunc()

print(f'{out1=}')
'''
out1=tensor([[44],
        [39],
        [33],
        [60]])
'''
print(f'{out2=}')
'''
out2=tensor([[33],
        [60],
        [63],
        [79]])
'''

torch.testing.assert_close(out1, out2, equal_nan=True)
```

Problem:
RNG state is inconsistent between eager execution and torch.compile(backend='inductor') execution, even after explicitly resetting the seed with torch.manual_seed(0) before each run.
The issue occurs when two randint calls are called inside the compiled function:the output of the first randint already differs between eager and compiled mode, even though the second randint result is not returned.

After a few debugging, I notice that the following version does NOT trigger the inconsistency:

```python
def fn():
    torch.manual_seed(0)
    out = torch.randint(0, 100, (4,1), dtype=torch.int64)
    torch.manual_seed(0)  # Extra seed reset before second randint
    other = torch.randint(0, 100, (2,1), dtype=torch.int64)
    return out
```

The only difference is an extra torch.manual_seed(0) before the second randint.
This behavior is bit counterintuitive: We already set the seed before the first `randint`, and the second `randint` does not modify the returned value.
However, without setting the second `randint`'s rng seed still causes the first randint output to mismatch between eager and compiled mode.

Note that this inconsistency only occurs for `inductor` backend, but does not occur for `eager/aot_eager`backend.

### Versions

[pip3] torch==2.11.0.dev20260209+cpu
[pip3] triton==3.6.0
[conda] torch 2.11.0.dev20260209+cpu pypi_0 pypi
[conda] triton 3.6.0 pypi_0 pypi

cc @ezyang @gchanan @kadeng @msaroufim @pbelevich @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by r83575** [comment]
I've opened a draft PR with a proposed fix: [https://github.com/pytorch/pytorch/pull/176842](https://github.com/pytorch/pytorch/pull/176842)

The change prevents RNG ops (such as `aten.randint`) from being reordered in `reorder_for_locality`,
which can otherwise lead to inconsistent RNG results between eager and compiled execution.

I would appreciate a response.

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260512+cu130`) on NVIDIA H100 but could not reproduce it (all 3 runs passed).

All 3 runs passed: eager and inductor-compiled outputs are identical ([[44],[39],[33],[60]]). torch.equal returned True every run.

<details>
<summary>Error output</summary>

```
out1=
tensor([[44],
        [39],
        [33],
        [60]])
out2=
tensor([[44],
        [39],
        [33],
        [60]])
equal: True

```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch
import torch._dynamo

torch._inductor.config.fallback_random = True


def _build_fn():
    def fn():
        torch.manual_seed(0)
        out = torch.randint(0, 100, (4, 1), dtype=torch.int64)
        other = torch.randint(0, 100, (2, 1), dtype=torch.int64)
        return out
    return fn


def test_fn() -> bool:
    """Return True if NO bug (eager == compiled). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")

    fn = _build_fn()
    cfunc = torch.compile(fn, backend=backend)

    torch.manual_seed(0)
    out1 = fn()
    torch.manual_seed(0)
    out2 = cfunc()

    return torch.equal(out1, out2)


if __name__ == "__main__":
    def fn():
        torch.manual_seed(0)
        out = torch.randint(0, 100, (4, 1), dtype=torch.int64)
        other = torch.randint(0, 100, (2, 1), dtype=torch.int64)
        return out

    cfunc = torch.compile(fn, backend="inductor")

    torch.manual_seed(0)
    out1 = fn()
    torch.manual_seed(0)
    out2 = cfunc()

    print(f"out1=\n{out1}")
    print(f"out2=\n{out2}")
    print(f"equal: {torch.equal(out1, out2)}")
    if not torch.equal(out1, out2):
        print(f"max abs diff: {(out1 - out2).abs().max().item()}")

    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260512+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260512+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260513+cu130
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.27.0.dev20260513+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.

**Comment by karthickai** [comment]
Closing this issue as Repro agent detected it was resolved in torch nightly; please reopen, if it arises again.

**Comment by r83575** [comment]
You're right, see the fix in this PR: [[inductor] Fix RNG order change caused by reorder_for_locality](https://github.com/pytorch/pytorch/pull/176842)
